vmm/sandbox/src/cgroup.rs

```diff
@@ -0,0 +1,425 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Ok, Result};
+use cgroups_rs::{
+    cgroup_builder::*, cpu::CpuController, cpuset::CpuSetController, hugetlb::HugeTlbController,
+    memory::MemController, Cgroup, Hierarchy, Subsystem,
+};
+use containerd_sandbox::{cri::api::v1::LinuxContainerResources, data::SandboxData};
+
+use crate::{
+    utils::{
+        get_overhead_resources, get_resources, get_sandbox_cgroup_parent_path, get_total_resources,
+    },
+    vm::VcpuThreads,
+};
+
+pub const VCPU_CGROUP_NAME: &str = "vcpu";
+pub const POD_OVERHEAD_CGROUP_NAME: &str = "pod_overhead";
+
+/// `SandboxCgroup` represents a set of cgroups for a sandbox.
+#[derive(Default, Debug)]
+pub struct SandboxCgroup {
+    pub sandbox_cgroup: Cgroup,
+    pub vcpu_cgroup: Cgroup,
+    pub pod_overhead_cgroup: Cgroup,
+}
+
+impl SandboxCgroup {
+    pub fn create_sandbox_cgroups(sandbox_data: &SandboxData) -> Result<Self> {
+        let sandbox_cgroup_parent_path = get_sandbox_cgroup_parent_path(sandbox_data)
+            .ok_or(anyhow!("failed to get sandbox cgroup parent path"))?;
+        let all_hiers = cgroups_rs::hierarchies::auto();
+
+        // Create sandbox cgroup in the all cgroup subsystem dir
+        let sandbox_cgroup_path_str = format!("{}/{}", sandbox_cgroup_parent_path, sandbox_data.id);
+        let mut sandbox_cgroup_path = Path::new(sandbox_cgroup_path_str.as_str());
+        // CgroupBuilder::new() func doesn't accept the cgroup name has "/" prefix,
+        // So need to remove the "/" prefix for sandbox_cgroup_path
+        if sandbox_cgroup_path.has_root() {
+            sandbox_cgroup_path = sandbox_cgroup_path.strip_prefix("/")?;
+        }
+
+        let sandbox_cgroup =
+            CgroupBuilder::new(sandbox_cgroup_path.to_str().unwrap()).build(all_hiers)?;
+
+        // Only create the vcpu and pod_overhead cgroups in the cpu cgroup subsystem
+        let vcpu_cgroup_path = format!(
+            "{}/{}",
+            sandbox_cgroup_path.to_str().unwrap(),
+            VCPU_CGROUP_NAME
+        );
+        let vcpu_cgroup =
+            CgroupBuilder::new(vcpu_cgroup_path.as_str()).build(cpu_subsystem_hier())?;
```

Comment:
`Cgroup::new_with_relative_paths()` may helps

Reply:
Fix it.